ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -55,61 +57,88 @@ def spin(self, rate=50):
                 self.update_waypoints(idx)
-            
+
             spin_rate.sleep()
 
 
     def pose_cb(self, pose):
         # TODO: Implement
-        self.__current_pose = pose.pose            
-        
+        self.__current_pose = pose.pose
+
 
     def waypoints_cb(self, lane):
         # Check out document for PoseStamped Message: http://docs.ros.org/melodic/api/geometry_msgs/html/msg/PoseStamped.html
         rospy.loginfo("Received base waypoints")
         if not self.__waypoints_tree:
             self.__base_waypoints = lane.waypoints
-            self.__base_waypoints_array = [[w.pose.pose.position.x, w.pose.pose.position.y] for w in self.__base_waypoints]            
-            self.__waypoints_tree = KDTree(self.__base_waypoints_array)                
-            rospy.loginfo("Successfully ingested based waypoints")                    
+            self.__base_waypoints_array = [[w.pose.pose.position.x, w.pose.pose.position.y] for w in self.__base_waypoints]
+            self.__waypoints_tree = KDTree(self.__base_waypoints_array)
+            rospy.loginfo("Successfully ingested based waypoints")
 
     def get_nearest_waypoint_id(self, pose):
         """
         Returns the nearest waypoint position index for the current pose
-        """        
+        """
         idx = self.__waypoints_tree.query([pose.position.x, pose.position.y])[1]
-        
+
         closest_point = self.__base_waypoints_array[idx]
         previous_point = self.__base_waypoints_array[idx - 1]
 
         closest_vector = np.array(closest_point)
         previous_vector = np.array(previous_point)
         current_pos_vector =  np.array([self.__current_pose.position.x, self.__current_pose.position.y])
 
-        
+
         val = np.dot(closest_vector - previous_vector, current_pos_vector - closest_vector)
         if val > 0:
             return (idx + 1) % len(self.__base_waypoints_array)
-        
+
         return idx
 
     def update_waypoints(self, idx):
         """
         Publishes the latest look-ahead waypoints
-        """        
+        """
         # Create the header and set its timestamp
         header = Header()
         header.stamp = rospy.Time.now()
-        
+
         msg = Lane()
         msg.header = header
-        msg.waypoints = self.__base_waypoints[idx: idx + LOOKAHEAD_WPS]
-
-        self.final_waypoints_pub.publish(msg)        
-
+        # Keep the copy of base_waypoints so that you don't have to recompute them
+        # we are using the same base_waypoints when we get multiple messages for stopping
+        # at a stopline.
+        base_waypoints = self.__base_waypoints[idx: idx + LOOKAHEAD_WPS]
+        msg.waypoints = base_waypoints
+        # If you find out that one of the generated waypoints lies on a stop line
+        # that we should be stopping at then start decelerating
+        if self.__stopline_wp_idx != -1 and self.__stopline_wp_idx < (idx + LOOKAHEAD_WPS):
```

Comment:
In python you could write:
```
idx < self.__stopline_wp_idx < idx + LOOKAHEAD_WPS
```

This is slightly more succinct than the above. Just a note on how we can make code even better. No need to change it now :+1: